### DISCUSSION TITLE
Is there a way we can use multiple connections (conn ids) in one dag?

### DISCUSSION THREAD

**Discussion by apinchuk1** [apache/airflow#34659] [General]
I am trying to run spark with s3 as a dag, I have created two connections:
  1. Connection to spark cluster 
  2. Connection to s3 that stores spark logs (history logs)
 
Is there an option to specify both of these in one dag?

**Comment by Taragolis** [comment]
Could you a bit clarify what you try to achieve. Because DAGs itself doesn't use connections, but Tasks (Operators) use it.

**Reply by apinchuk** [reply]
You are right, the connection in the task. I am trying to use SparkSubmitOperator for example as follows:

with airflow.DAG('sparkTestDag',
                  default_args=default_args,
                  schedule_interval='0 1 * * *') as dag:

    spark_task = SparkSubmitOperator(
        task_id='spark_job1',
        conn_id='spark',
        application="sparkJob.py",
        conf={
              spark.eventLog.enabled=true,
              spark.eventLog.dir=s3a://spark-history-log-path/,
              spark.history.fs.logDirectory=s3a://spark-history-log-path/,
              spark.hadoop.fs.s3a.access.key=ACCESSKEY,
              spark.hadoop.fs.s3a.secret.key=SECRETKEY,
             }

Is there a way to omit specifying s3 connection details in the configuration of the spark task, but instead to call another "conn_id" where s3 connection is already established.

**Reply by Taragolis** [reply]
You could use connection as part of templates, more info you could found in [Airflow Connections in Templates](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#airflow-connections-in-templates), however it would only works if you stored your permanent credentials into login (AWS Access Key ID) and password (AWS Secret Access Key):

```python
spark_task = SparkSubmitOperator(
    task_id='spark_job1',
    conn_id='spark',
    application="sparkJob.py",
    conf={
          ...
          "spark.hadoop.fs.s3a.access.key": "{{ conn.your_aws_conn_id.login }}",
          "spark.hadoop.fs.s3a.secret.key": "{{ conn.your_aws_conn_id.password }}"
    }
    ...
)
```

**Reply by apinchuk** [reply]
Perfect, this is exactly what I was looking for. Thank you!
